man/usernetctl.8
.TH USERNETCTL 8 "Red Hat, Inc." "RHS" \" -*- nroff -*-
.SH NAME
usernetctl \- allow a user to manipulate a network interface if permitted
.SH SYNOPSIS
.B usernetctl
\fIinterface-name\fP up\fI|\fPdown\fI|\fPreport
.SH DESCRIPTION
.B usernetctl
checks to see if users are allowed to manipulate the network interface
specified by \fIinterface-name\fP, and then tries to bring the network
interface up or down, if up or down was specified on the command line,
or returns true or false status (respectively) if the report option was
specified.

.B usernetctl
is not really meant to be called directly by users, though it currently
works fine that way.  It is used as a wrapper by the ifup and ifdown
scripts, so that users can do exactly the same thing as root:
.nf
ifup \fIinterface-name\fP
ifdown \fIinterface-name\fP
.fi
and \fBifup\fP and \fBifdown\fP will call usernetctl automatically to
allow the interface status change.
.SH OPTIONS
.TP
.I "\fIinterface-name"
The name of the network interface to check; for example, "ppp0".  For
backwards compatibility, "ifcfg-ppp0" and
"/etc/sysconfig/network-scripts/ifcfg-ppp0" are also supported.
.TP
up\fI|\fPdown
Attempt to bring the interface up or down.
.TP
report
Report on whether users can bring the interface up or down.
.SH NOTES
Alternate device configurations may inherit the default configuration's
permissions.


ChangeLog
commit f8c758709c2dea6651afcc75fa7cceb3b7ff7afc
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 18:17:22 2010 +0200

    version 8.45.32

 initscripts.spec |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 967b64ce3f9754f7de902ed290c813634c182803
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 18:16:21 2010 +0200

    52a7a1f115bb87323620b73f5dde21e2c5e7d9a7

 ChangeLog                        |    7 ++++---
 udev/rules.d/54-cu3088-fix.rules |   10 ++++++++++
 2 files changed, 14 insertions(+), 3 deletions(-)

commit c014f9c6c09b50302e89f9df21077343b6c363b8
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 17:34:33 2010 +0200

    initrscripts.spec: sync with cvs repo

 initscripts.spec |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 923ef28cc4d4654e58033c81cd44ce9e1e4734b8
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 17:30:23 2010 +0200

    version 8.45.31

 ChangeLog        | 1016 ++++++++++++++++++++++++++++++++++--------------------
 initscripts.spec |    4 +-
 2 files changed, 636 insertions(+), 384 deletions(-)

commit 52a7a1f115bb87323620b73f5dde21e2c5e7d9a7
Author: Phil Knirsch <pknirsch@redhat.com>
Date:   Mon Sep 13 16:29:45 2010 +0200

    s390: Added required 54-cu3088-fix.rules udev rules file (#475721)

 initscripts.spec |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

commit b74c943aa01f6a8facc36246dc20ccebd9eb690c
Author: David Cantrell <dcantrell@redhat.com>
Date:   Fri Sep 26 16:41:21 2008 -1000

    Expand get_hostname() function to support IPv4 and IPv6.
    
    Make sure that get_hostname() if the user specifies either address
    family.

 src/ipcalc.c |   36 +++++++++++++++++++++++++++++-------
 1 files changed, 29 insertions(+), 7 deletions(-)

commit f085cdfdc0d67fe240c23e75d14dce6534d36c8a
Author: David Cantrell <dcantrell@redhat.com>
Date:   Fri Sep 26 16:30:38 2008 -1000

    Remove unnecessary tab.

 src/ipcalc.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit e921ac88353159dee451f7c0094fca95a9b2bb56
Author: David Cantrell <dcantrell@redhat.com>
Date:   Fri Sep 26 16:29:21 2008 -1000

    Correct address family option handling in ipcalc.
    
    Make sure the default is IPv4.  If the check option is specified,
    make sure we just validate the address and return 1 or 0.

 src/ipcalc.c |   83 ++++++++++++++++++++++++++++-----------------------------
 1 files changed, 41 insertions(+), 42 deletions(-)

commit 9253ed4bdddd635706a0a741259cbaa8a21d8b89
Author: David Cantrell <dcantrell@redhat.com>
Date:   Fri Sep 26 16:26:27 2008 -1000

    Compile ipcalc with _GNU_SOURCE defined (for asprintf).

 src/Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 192392a47594d8141aaf14a0a705bc421899adb0
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 16:20:36 2010 +0200

    remove deprecation comments for raw devices and fix path

 initscripts.spec       |    6 ++++++
 rc.d/init.d/rawdevices |    6 ++----
 sysconfig/rawdevices   |    3 ---
 3 files changed, 8 insertions(+), 7 deletions(-)

commit 6c0c324a6c0c031ed4604c6a020829503e8d086f
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 15:59:40 2010 +0200

    version 8.45.30

 initscripts.spec |   28 +++++++++++++++++++++++++++-
 1 files changed, 27 insertions(+), 1 deletions(-)

commit 8f65b61fc53d245c7865bfa0ab879337d9c6ff31
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 15:48:34 2010 +0200

    rc.sysinit: also remove /var/run/dovecot/*
    
    https://bugzilla.redhat.com/show_bug.cgi?id=492994

 rc.d/rc.sysinit |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 85f809ad1d6f47d4fe73a07fa358fd917bc94b87
Author: Bill Nottingham <notting@redhat.com>
Date:   Wed Sep 2 16:28:02 2009 -0400

    Doucment DHCP_HOSTNAME (#498052)

 sysconfig.txt |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

commit 52dfb70cd860e4d0fd51e37e30a081fd197f1b74
Author: Bill Nottingham <notting@redhat.com>
Date:   Fri Jul 20 20:54:47 2007 +0000

    move quotacheck to after filesystem mounting so it actually works (#249003, <tometzky@batory.org.pl>)
    also, nuke the convertquota stuff that has been obsolete (and not working) for quite some time

 rc.d/rc.sysinit |   25 +++++--------------------
 1 files changed, 5 insertions(+), 20 deletions(-)

commit 873660981a5a42a6de411e04958a4f204704efe1
Author: Bill Nottingham <notting@redhat.com>
Date:   Tue Dec 15 10:59:44 2009 -0500

    Silence error. (#516569)

 sysconfig/network-scripts/network-functions |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 3ae2380e43092dfdc12b2d242ade6bd1379fe1e7
Author: Bill Nottingham <notting@redhat.com>
Date:   Tue Oct 20 16:56:32 2009 -0400

    Load bonding driver if BONDING_OPTS is defined. (#516569)

 sysconfig/network-scripts/network-functions |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

commit f08000519addd0b13fada0eba4afef72871613c1
Author: Bill Nottingham <notting@redhat.com>
Date:   Thu Aug 27 13:25:41 2009 -0400

    Restore context of recreated files. (#519748)

 rc.d/rc.sysinit |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

commit 209e5e18cbd1cbb76f7e49c83854763c4e9f1d1c
Author: Bill Nottingham <notting@redhat.com>
Date:   Fri Oct 9 16:18:04 2009 -0400

    Do not try to activate subsets and layer raid and handle 'no raid disks'.
    
    - 'no raid disks' handling: (#527726, <mschmidt@redhat.com>)
    - no subsets: https://bugzilla.redhat.com/show_bug.cgi?id=524356

 rc.d/rc.sysinit |   15 +++++++--------
 1 files changed, 7 insertions(+), 8 deletions(-)

commit 40ddde4c4e6ad1dc855a8a3f4ae594d6c2d16179
Author: Bill Nottingham <notting@redhat.com>
Date:   Wed Oct 7 11:10:12 2009 -0400

    Add /var/lib/xend. (#526046)

 rwtab |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

commit 643c05c59a581663140e307f08732c6bef3d7622
Author: Bill Nottingham <notting@redhat.com>

src/usernetctl.c
/*
 * Copyright (c) 1997-2007 Red Hat, Inc. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

#include <alloca.h>
#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>
#include <limits.h>

/* This will be running setuid root, so be careful! */
static const char * safeEnviron[] = {
	"PATH=/bin:/sbin:/usr/bin:/usr/sbin",
	"HOME=/root",
	NULL
};

#define FOUND_FALSE -1
#define NOT_FOUND 0
#define FOUND_TRUE 1

static void
usage(void) {
    fprintf(stderr, "usage: usernetctl <interface-config> <up|down|report>\n");
    exit(1);
}

static size_t
testSafe(char *ifaceConfig, int fd) {
    struct stat sb;

    /* These shouldn't be symbolic links -- anal, but that's fine w/ mkj. */
    if (fstat(fd, &sb)) {
	fprintf(stderr, "failed to stat %s: %s\n", ifaceConfig, 
		strerror(errno));
	exit(1);
    }

    /* Safety/sanity checks. */
    if (!S_ISREG(sb.st_mode)) {
	fprintf(stderr, "%s is not a normal file\n", ifaceConfig);
	exit(1);
    }

    if (sb.st_uid) {
	fprintf(stderr, "%s should be owned by root\n", ifaceConfig);
	exit(1);
    }
    
    if (sb.st_mode & S_IWOTH) {
	fprintf(stderr, "%s should not be world writeable\n", ifaceConfig);
	exit(1);
    }

    return sb.st_size;
}


static int
userCtl(char *file) {
    char *buf;
    char *contents = NULL;
    char *chptr = NULL;
    char *next = NULL;
    int fd = -1, retval = NOT_FOUND;
    size_t size = 0;

    /* Open the file and then test it to see if we like it. This way
       we avoid switcheroo attacks. */
    if ((fd = open(file, O_RDONLY)) == -1) {
	fprintf(stderr, "failed to open %s: %s\n", file, strerror(errno));
	exit(1);
    }

    size = testSafe(file, fd);
    if (size > INT_MAX) {
        fprintf(stderr, "file %s is too big\n", file);
        exit(1);
    }

    buf = contents = malloc(size + 2);
    if (contents == NULL) {
        fprintf(stderr, "failed to allocate memory\n");
        exit(1);
    }

    if (read(fd, contents, size) != size) {
	perror("error reading device configuration");
	exit(1);
    }
    close(fd);

    contents[size] = '\n';
    contents[size + 1] = '\0';

    /* Each pass parses a single line (until an answer is found),  The contents
       pointer itself points to the beginning of the current line. */
    while (*contents) {
	chptr = contents;
	while (*chptr != '\n') chptr++;
	next = chptr + 1;
	while (chptr >= contents && isspace(*chptr)) chptr--;
	*(++chptr) = '\0';

	if (!strncasecmp(contents, "USERCTL=", 8)) {
	    contents += 8;
	    if ((contents[0] == '"' &&
		 contents[strlen(contents) - 1] == '"') ||
		(contents[0] == '\'' &&
		 contents[strlen(contents) - 1] == '\''))
		{
		contents++;
		contents[strlen(contents) - 1] = '\0';
	    }

	    if (!strcasecmp(contents, "yes") || !strcasecmp(contents, "true")) 
		retval = FOUND_TRUE;
	    else 
		retval = FOUND_FALSE;

	    break;
	}

	contents = next;
    }

    free(buf);

    return retval;
}

int
main(int argc, char ** argv) {
    char * ifaceConfig;
    char * chptr;
    char * cmd = NULL;
    int report = 0;
    char tmp;

    if (argc != 3) usage();

    if (!strcmp(argv[2], "up")) {
	cmd = "./ifup";
    } else if (!strcmp(argv[2], "down")) {
	cmd = "./ifdown";
    } else if (!strcmp(argv[2], "report")) {
	report = 1;
    } else {
	usage();
    }

    if (chdir("/etc/sysconfig/network-scripts")) {
	fprintf(stderr, "error switching to /etc/sysconfig/network-scripts: "
		"%s\n", strerror(errno));
	exit(1);
    }

    /* force the interface configuration to be in the current directory */
    chptr = ifaceConfig = argv[1];
    while (*chptr) {
	if (*chptr == '/')
	    ifaceConfig = chptr + 1;
	chptr++;
    }

    /* automatically prepend "ifcfg-" if it is not specified */
    if (strncmp(ifaceConfig, "ifcfg-", 6)) {
	char *temp;
        size_t len = strlen(ifaceConfig);

	/* Make sure a wise guys hasn't tried an integer wrap-around or
	   stack overflow attack. There's no way it could refer to anything 
	   bigger than the largest filename, so cut 'em off there. */
        if (len > PATH_MAX)
		exit(1);

	temp = (char *) alloca(len + 7);
	strcpy(temp, "ifcfg-");
	/* strcat is safe because we got the length from strlen */
	strcat(temp, ifaceConfig);
	ifaceConfig = temp;
    }
    
    if(getuid() != 0)
    switch (userCtl(ifaceConfig)) {
	char *dash;

	case NOT_FOUND:
	    /* a `-' will be found at least in "ifcfg-" */
	    dash = strrchr(ifaceConfig, '-');
	    if (*(dash-1) != 'g') {
		/* This was a clone configuration; ask the parent config */
		tmp = *dash;
		*dash = '\0';
		if (userCtl(ifaceConfig) == FOUND_TRUE) {
		    /* exit the switch; users are allowed to control */
		    *dash = tmp;
		    break;
		}
		*dash = tmp;
	    }
	    /* else fall through */
	case FOUND_FALSE:
	    if (! report)
	        fprintf(stderr,
			"Users are not allowed to control this interface.\n");
	    exit(1);
	    break;
    }

    /* looks good to me -- let's go for it if we are changing the interface,
     * report good status to the user otherwise */

    if (report)
	exit(0);

    /* pppd wants the real uid to be the same as the effective (god only
       knows why when it works fine setuid out of the box) */
    setuid(geteuid());
    /* Drop user gid (for temp files, SELinux) */
    setgid(0);

    execle(cmd, cmd, ifaceConfig, NULL, safeEnviron);
    fprintf(stderr, "exec of %s failed: %s\n", cmd, strerror(errno));
    
    exit(1);
}
